src/darwin/Framework/CHIP/MTRDeviceController.mm

```diff
@@ -311,7 +320,53 @@ - (BOOL)isRunning
     return _cppCommissioner != nullptr;
 }
 
+- (NSUInteger)shutdownPrecheck
+{
+    __block NSUInteger assertionCount = 0;
+    dispatch_sync(_chipWorkQueue, ^{
+        self.shutdownPending = TRUE;
+        assertionCount = self.keepRunningAssertionCounter;
+    });
+    return assertionCount;
+}
+
+- (void)addRunAssertion
+{
+    dispatch_sync(_chipWorkQueue, ^{
+        ++self.keepRunningAssertionCounter;
+        MTR_LOG("%@ Adding keep running assertion, total %lu", self, (unsigned long) self.keepRunningAssertionCounter);
+    });
+}
+
+- (void)removeRunAssertion;
+{
+    __block BOOL shouldShutdown = FALSE;
+    dispatch_sync(_chipWorkQueue, ^{
+        if (self.keepRunningAssertionCounter > 0) {
+            --self.keepRunningAssertionCounter;
+            MTR_LOG("%@ Removing keep running assertion, total %lu", self, (unsigned long) self.keepRunningAssertionCounter);
+            if (self.keepRunningAssertionCounter == 0 && self.shutdownPending) {
+                shouldShutdown = TRUE;
+            }
+        }
+    });
+    if (shouldShutdown) {
+        MTR_LOG("%@ All assertions removed and shutdown is pending, shutting down", self);
+        [self finalShutdown];
+    }
+}
+
 - (void)shutdown
+{
+    NSUInteger assertionCount = [self shutdownPrecheck];
```

Comment:
So I think there is a race here: what if one thread calls shutdown while another is calling addRunAssertion?

We could not worry about that race, I guess, or we could make addRunAssertion fallible and have it fail if we are in a situation where shutdownPrecheck has already returned 0 (would have to set that state under the lock or on the matter queue in shutdownPrecheck).